src/v2/Artsy/Relay/middleware/cache/Cache.ts

```diff
@@ -0,0 +1,102 @@
+import { QueryResponseCache } from "relay-runtime"
+import createLogger from "v2/Utils/logger"
+import RelayQueryResponseCache from "relay-runtime/lib/network/RelayQueryResponseCache"
+import { isServer } from "lib/environment"
+
+const logger = createLogger("v2/Artsy/middleware/cache/Cache")
+
+export interface CacheConfig {
+  size: number
+  ttl: number
+}
+
+export class Cache {
+  cacheConfig: CacheConfig
+  relayCache: RelayQueryResponseCache
+
+  redisCache: {
+    del: (key: string) => Promise<void>
+    expire: (key: string, ttl: number) => Promise<void>
+    flushall: () => Promise<void>
+    get: (key: string) => Promise<string>
+    set: (key: string, value: string) => Promise<void>
+  }
+
+  constructor(cacheConfig: CacheConfig) {
+    this.cacheConfig = cacheConfig
+    this.initRelayCache()
+    this.initRedisCache()
+  }
+
+  initRelayCache() {
+    this.relayCache = new QueryResponseCache(this.cacheConfig)
+  }
+
+  initRedisCache() {
+    if (!isServer) {
+      return
+    }
+
+    const { promisify } = require("util")
+    const redis = require("redis")
+    const client = redis.createClient()
+
+    this.redisCache = {
+      del: promisify(client.del).bind(client),
+      expire: promisify(client.expire).bind(client),
+      flushall: promisify(client.flushall).bind(client),
+      get: promisify(client.get).bind(client),
+      set: promisify(client.set).bind(client),
+    }
+  }
+
+  getCacheKey(query, variables) {
+    const cacheKey = JSON.stringify({ query, variables })
+    return cacheKey
+  }
+
+  async get(queryId, variables) {
+    let cachedRes = this.relayCache.get(queryId, variables)
+
+    // No cache in relay store, check redis
+    if (isServer && !cachedRes) {
+      const cacheKey = this.getCacheKey(queryId, variables)
+
+      try {
+        cachedRes = JSON.parse(await this.redisCache.get(cacheKey))
+        if (cachedRes) {
+          logger.log("\nCache operation: [get]", cacheKey)
+        }
+      } catch (error) {
+        logger.error("Error retrieving cache: [get]", cacheKey, error)
+      }
+    }
+
+    return cachedRes
+  }
+
+  async set(queryId, variables, res) {
+    this.relayCache.set(queryId, variables, res)
+
+    // Store in redis during server-side pass
+    if (isServer) {
+      const cacheKey = this.getCacheKey(queryId, variables)
+
+      try {
+        await this.redisCache.set(cacheKey, JSON.stringify(res))
+        await this.redisCache.expire(cacheKey, this.cacheConfig.ttl)
```

Comment:
For server-side expiration, we can also make this more configurable. Right now it's piggybacking on the same values currently used on the client.